sdk-development-guidelines.md

```diff
@@ -0,0 +1,200 @@
+* SHOULD use quality tools available in common config: [Android](https://github.com/rakutentech/android-buildconfig) and [iOS](https://github.com/rakutentech/ios-buildconfig)
+* For Android <img src="images/android.png" alt="drawing" width="15"/>:
+    * MUST use the [Android Support Annotations](https://developer.android.com/studio/write/annotations) for public APIs
+    * Non-Android Java code MUST use the [Jetbrains Annotations](https://www.jetbrains.com/help/idea/annotating-source-code.html) for public APIs
+
+### Library Design
+
+#### Limit impact on consumers
+* SHOULD limit dependencies as much as possible primarily due to risk of version clashes e.g. SHOULD NOT depend on external networking library but also because the dependency may stop being maintained
+* SHOULD NOT have exported dependency on other SDK modules
+* SHOULD NOT have an exported dependency on other SDK modules e.g. if your module needs authentication, do not depend on Authentication SDK. Instead, require your library consumer to pass an access token string to your API instead.
+* SHOULD NOT have unnecessary exported dependencies, in particular
+    * SHOULD NOT leak networking library to consumers
+    * SHOULD NOT leak serialization library to consumers
+    * An exception to this rule are optional adapters for popular libraries
+* For Android <img src="images/android.png" alt="drawing" width="15"/>:
+    * MUST use a [consumerProguardFiles](https://google.github.io/android-gradle-dsl/current/com.android.build.gradle.internal.dsl.BuildType.html#com.android.build.gradle.internal.dsl.BuildType:consumerProguardFiles) file if obfuscation breaks the libraries function
+
+#### Expose minimal Surface area
+* Keep everything private unless it needs to be public. Every public class, method, property & resource MUST have a strong reason for being public
+* MAY keep internal source code in a `/Private` folder to allow tools to more easily ignore those files
+* MUST hide classes that need to be public for technical reasons but are not intended for SDK clients from documentation
+* SHOULD declare classes as final to avoid unexpected high-jacking of SDK behavior by subclasses
+* For Android <img src="images/android.png" alt="drawing" width="15"/>:
+    * SHOULD use `RestrictTo` annotation on classes that are internal to the library
+
+#### Best Practices
+* Simple configuration keys MAY be settable in application's manifest file (Android) or Info.plist (iOS) or passed directly to an API configuration function:
+* For Android <img src="images/android.png" alt="drawing" width="15"/>:
+    * the manifest file is read in a [(fake) ContentProvider](https://firebase.googleblog.com/2016/12/how-does-firebase-initialize-on-android.html) and be automatically initialized (if that fits with your library design)
+    * SHOULD NOT require manual initialization such as by overriding the `Application` class or depending on a host app trigger
```

Comment:
IAM depends on a host app trigger so I'm not sure we should include this guideline.

Reply:
removed